From 8a25769817487c22b6f9edda33a736656ca6bb42 Mon Sep 17 00:00:00 2001
From: Yangbo Lu <yangbo.lu@nxp.com>
Date: Mon, 4 Mar 2019 12:04:14 +0800
Subject: [PATCH 657/706] enetc: read TX timestamp from writeback descriptor

TX timestamp should be read from writeback descriptor not
from the extension BD, even they are using similar format.
This patch is to define transmit buffer writeback descriptor,
and to use it for TX timestamp getting.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>
Signed-off-by: Claudiu Manoil <claudiu.manoil@nxp.com>
(cherry picked from commit d0156f8bb55467f41c50ab7235496e8f493a9a36)
Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 drivers/net/ethernet/freescale/enetc/enetc.c    | 4 ++--
 drivers/net/ethernet/freescale/enetc/enetc_hw.h | 6 ++++++
 2 files changed, 8 insertions(+), 2 deletions(-)

diff --git a/drivers/net/ethernet/freescale/enetc/enetc.c b/drivers/net/ethernet/freescale/enetc/enetc.c
index 86e40e61d786..2f22e3bf7b00 100644
--- a/drivers/net/ethernet/freescale/enetc/enetc.c
+++ b/drivers/net/ethernet/freescale/enetc/enetc.c
@@ -308,9 +308,9 @@ static void enetc_get_tx_tstamp(struct enetc_hw *hw, union enetc_tx_bd *txbd,
 	if (txbd->flags & ENETC_TXBD_FLAGS_W) {
 		lo = enetc_rd(hw, ENETC_SICTR0);
 		hi = enetc_rd(hw, ENETC_SICTR1);
-		if (lo <= txbd->ext.tstamp)
+		if (lo <= txbd->wb.tstamp)
 			hi -= 1;
-		*tstamp = (u64)hi << 32 | txbd->ext.tstamp;
+		*tstamp = (u64)hi << 32 | txbd->wb.tstamp;
 	}
 }
 
diff --git a/drivers/net/ethernet/freescale/enetc/enetc_hw.h b/drivers/net/ethernet/freescale/enetc/enetc_hw.h
index 8cb37f777c20..c57ddede4d89 100644
--- a/drivers/net/ethernet/freescale/enetc/enetc_hw.h
+++ b/drivers/net/ethernet/freescale/enetc/enetc_hw.h
@@ -362,6 +362,12 @@ union enetc_tx_bd {
 		u8 e_flags;
 		u8 flags;
 	} ext; /* Tx BD extension */
+	struct {
+		__le32 tstamp;
+		u8 reserved[10];
+		u8 status;
+		u8 flags;
+	} wb; /* writeback descriptor */
 };
 
 #define ENETC_TXBD_FLAGS_L4CS	BIT(0)
-- 
2.17.1

